ios/brave-ios/Sources/BraveShared/AsyncFileManager.swift

```diff
@@ -0,0 +1,335 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+/// A drop-in replacement for FileManager which guarantees at compile time that operations execute
+/// on the cooperative thread pool and not the UI thread. Reach for this by default unless file i/o
+/// is required to be done in sequence with other calls (at which point, use FileManager + Dispatch)
+///
+/// Simply replace `FileManager.default` with `AsyncFileManager.default` and add `await`
+public final class AsyncFileManager {
+  public static let `default` = AsyncFileManager(fileManager: .default)
+
+  /// The underlying file manager to execute items on
+  private let fileManager: FileManager
+
+  /// Create an `AsyncFileManager` with a `FileManager`
+  public init(fileManager: FileManager) {
+    self.fileManager = fileManager
+  }
+
+  // MARK: - Accessing User Directories
+
+  /// Locates the specified common directory in a domain.
+  public func url(
+    for directory: FileManager.SearchPathDirectory,
+    in domain: FileManager.SearchPathDomainMask
+  ) throws -> URL {
+    // This method is non-async as `create` is explicitly passed as `false`
+    try fileManager.url(for: directory, in: domain, appropriateFor: nil, create: false)
+  }
+
+  /// Locates and optionally creates the specified common directory in a domain.
+  public func url(
+    for directory: FileManager.SearchPathDirectory,
+    in domain: FileManager.SearchPathDomainMask,
+    create: Bool
+  ) async throws -> URL {
```

Comment:
I don't understand some of these methods. Why is this async but not calling anything async internally?

Reply:
They are all marked async because then they will run in a nonisolated context (non-main)

Reply:
Ahh missed that